keanu-project/src/main/java/io/improbable/keanu/vertices/VertexLabel.java

```diff
@@ -1,30 +1,58 @@
 package io.improbable.keanu.vertices;
 
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
 import java.util.Objects;
 
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableList;
+
 public class VertexLabel {
-    private final String label;
+    private static final char NAMESPACE_SEPARATOR = '.';
+    private static final char NAME_PREPENDER = ':';
+    private final String name;
+    private final List<String> namespace;
+
+    public VertexLabel(String name, String... namespace) {
+        this.name = name;
+        this.namespace = ImmutableList.copyOf(namespace);
+    }
+
+    public VertexLabel inNamespace(String topLevelNamespace) {
```

Comment:
I see what this does, but I wonder if the name is slightly confusing (when doing my first skim read through I thought this was some form of matching function rather than something that put on a new layer of name-spacing).

Reply:
Got any suggestions for a better name?

Reply:
A tricky one :)

perhaps addNamespace or plusNamespace or similar?

Reply:
Or encapsulate()?

Reply:
In order to indicate that they are non-mutating, how about `withExtraNamespace(String)` and `withoutOuterNamespace()`?